hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -571,18 +571,24 @@ TcpClientConnection getOrConnect(@Nonnull Address address) {
             return connection;
         }
 
-        synchronized (resolveAddress(address)) {
-            // this critical section is used for making a single connection
-            // attempt to the given address at a time.
-            connection = getConnection(address);
-            if (connection != null) {
+        InetSocketAddress inetSocketAddress = resolveAddress(address);
+        try {
+            Object mutex = ConcurrencyUtil.getOrPutIfAbsent(mutexes, inetSocketAddress, key -> new Object());
```

Comment:
I had a closer look at the code and this is on the ConnectionManager. I don't see a problem if another thread blocks on acquring a concurrent connection. He can't do anything useful anyway since the method itself is blocking.

So using a lock isn't that much of a problem. 

I would be careful with using an intrinsic lock due to inflation/gc problems as I already indicated in my first comment. It will not immediately cause problems since connections will not be created and destroyed in a very high frequency. But eventuall you could still run into some GC issues.